### ISSUE TITLE
C system types are exposed, which may be incompatible

### ISSUE THREAD

**Issue by compnerd** [apple/swift-nio#1673]
The C system types are not guaranteed to be identical across all systems.  Windows exposes this issue when trying to build all the public interfaces which use the system defined interfaces.  The following (potentially incomplete) list of interfaces are the ones that currently show up:

- sockaddr_un
- sockaddr_in6
- sockaddr_in
- sockaddr
- in_addr
- linger

**Comment by compnerd** [comment]
I suppose that adding in `tcp_info` and `tcp_connection_info` to the list would be nice as well.  I'd like to expose `TCP_INFO_V0` so that Windows can also query the additional information about the TCP connection.

**Comment by Lukasa** [comment]
Thanks for this! In general these types are going to fall into a few categories. Some types may even be in more than one!

1. Things that need a proper Swift type.

    In this category is, for example, `in_addr` and `in6_addr` (#1650). These types desperately want a better, pure-Swift replacement that can have some semantic behaviours, rather than their fairly useless C-level counterparts. For things like this we should file issues to define a proper type. All of these also want to expose the current type using (2).

2. Things that exist as a point to extract the type for use with system APIs.

    In this category is, for example, `sockaddr_*`. These types exist primarily for transforming between the pure Swift representation (`SocketAddress`) and the equivalent low-level representation on the platform. For these types, I believe it's acceptable to have the system representation be different on different systems. We should consider deprecating existing access points for those types and replace them with `.systemRepresentation`.

3. Places where the API surface accepts a C type, where the API is low-level and platform specific.

    Here I'd say we see things like `linger`, `tcp_info`, and `tcp_connection_info`. These types are platform specific: their specificity is often revealed by the fact that they're already behind compiler guards (such as the two tcp info types). These APIs particularly commonly appear in low-level interfaces such as `SocketOptionProvider`. In these instances we can compile that API surface out on platforms where those types don't exist. Where possible we should investigate whether _similar_ APIs exist, as those should be added at the same time.

@weissi does this sound like a reasonable division of work to you?

**Comment by compnerd** [comment]
I don't think that the `linger` type is behind guards.  That said, its just a matter of exposing the typealias, and the structure itself is opaque, so its not _terrible_ but, I would rather discuss that before making the types public (which I've tried to avoid as much as possible).

**Comment by Lukasa** [comment]
Sorry, I wasn't saying linger was behind a guard but that it potentially _should_ be, if no equivalent socket operation existed on Windows.

**Comment by compnerd** [comment]
I think that's where the difference in the view is.  I think it would be best if we avoided leaking platform types wherever possible.  In the case of linger, since it's an opaque pointer, I don't think it matters too much, but anywhere where the user can interact with the type, we should wrap to help ensure that writing against NIO gives you some amount of independence from the underlying platform.  I think that we are in agreement though.

**Comment by Lukasa** [comment]
> I think that's where the difference in the view is. I think it would be best if we avoided leaking platform types wherever possible.

I mean, we can do that, but at a certain point we're just lying to ourselves. NIO defining `struct Linger` that looks exactly like C's `struct { } linger` is not really helping anyone achieve platform independence, not least because there's nothing we can do to compute on it.

In my view NIO should be conceptualising of things in layers. There is a low-level system layer where we allow users to escape the shackles of NIO's functionality and tell the system to do stuff. The socket options level is a good example of this: you can throw nearly anything in there and we'll let you do it. From time to time we provide helpers for commonly used bits of functionality there.

As things are more prominently used (e.g. `in_addr`) those things become good candidates for being rewritten in Swift, not so much because it makes them platform agnostic but because it makes them easier to use.

**Comment by MahdiBM** [comment]
Since @compnerd is not in the loop, I thought it's worth updating this issue and mention I've created the following issue which should resolve the IP-Address-related concerns of this issue, if/when it goes through:

https://github.com/apple/swift-nio/issues/3448 ("Use swift-native "SocketAddress" facilities")
